lib/ocpp/common/websocket/websocket_tls_tpm.cpp

```diff
@@ -398,24 +349,65 @@ void WebsocketTlsTPM::client_loop() {
 
     info.fd_limit_per_thread = 1 + 1 + 1;
 
-    bool use_tpm = connection_options.use_tpm_tls;
+    // Setup context - need to know the key type first
+
+    std::string path_key;
+    std::string path_chain;
+    std::optional<std::string> password;
 
-    // Setup context
-    OSSL_LIB_CTX* lib_ctx;
-    SSL_CTX* ssl_ctx;
+    if (this->connection_options.security_profile == 3) {
 
-    create_sec_context(use_tpm, lib_ctx, ssl_ctx);
+        const auto certificate_key_pair =
+            this->evse_security->get_key_pair(CertificateSigningUseEnum::ChargingStationCertificate);
 
-    // Connection acquire the contexts
-    conn_data->sec_lib_context = std::unique_ptr<OSSL_LIB_CTX>(lib_ctx);
-    conn_data->sec_context = std::unique_ptr<SSL_CTX>(ssl_ctx);
+        if (!certificate_key_pair.has_value()) {
+            EVLOG_AND_THROW(std::runtime_error(
+                "Connecting with security profile 3 but no client side certificate is present or valid"));
+        }
+
+        path_chain = std::string(certificate_key_pair.value().certificate_path.c_str());
+        path_key = std::string(certificate_key_pair.value().key_path.c_str());
+        password = certificate_key_pair.value().password;
+    }
+
+    SSL_CTX* ssl_ctx = nullptr;
+    bool tpm_key = false;
+
+    {
+        OpenSSLProvider provider;
+
+        if (!path_key.empty()) {
+            tpm_key = is_tpm_key_filename(path_key.c_str());
+#ifdef DEBUG
+            EVLOG_info << "TPM Key: " << tpm_key;
+#endif
+        }
+
+        if (tpm_key) {
+            provider.set_tls_mode(OpenSSLProvider::mode_t::tpm2_provider);
+        } else {
+            provider.set_tls_mode(OpenSSLProvider::mode_t::default_provider);
+        }
+
+        EVLOG_info << "Using TLS propquery: " << provider.propquery_tls_str();
+        const SSL_METHOD* method = SSLv23_client_method();
+        ssl_ctx = SSL_CTX_new_ex(provider, provider.propquery_tls_str(), method);
+
+        if (ssl_ctx == nullptr) {
+            ERR_print_errors_fp(stderr);
+            EVLOG_AND_THROW(std::runtime_error("Unable to create ssl ctx."));
+        }
+    }
 
     // Init TLS data
-    tls_init();
+    tls_init(ssl_ctx, path_chain.c_str(), path_key.c_str(), tpm_key, password);
```

Comment:
How about changing the function signature so you can pass a const string& here
```suggestion
    tls_init(ssl_ctx, path_chain, path_key, tpm_key, password);
```

Reply:
updated